File: .tmp_py4013/batch_17_dump/3891_minimum_increase_to_maximize_special_indices.js
// LeetCode 3891 - Minimum Increase To Maximize Special Indices
// https://leetcode.com/problems/minimum-increase-to-maximize-special-indices/

var minIncrease = function(nums) {
    const n = nums.length;
    const f = Array.from({length: n}, () => [-1, -1]);
    const dfs = (i, j) => {
        if (i >= n - 1) return 0;
        if (f[i][j] !== -1) return f[i][j];
        const cost = Math.max(0, Math.max(nums[i - 1], nums[i + 1]) + 1 - nums[i]);
        let ans = cost + dfs(i + 2, j);
        if (j > 0) ans = Math.min(ans, dfs(i + 1, 0));
        return f[i][j] = ans;
    };
    return dfs(1, (n & 1) ^ 1);
};

File: .tmp_rs4013/rust_check21/3891_minimum_increase_to_maximize_special_indices.rs
struct Solution;
// LeetCode 3891 - Minimum Increase to Maximize Special Indices
// https://leetcode.com/problems/minimum-increase-to-maximize-special-indices/

impl Solution {
    pub fn min_increase(nums: Vec<i32>) -> i64 {
        let n = nums.len();
        let mut f = vec![[-1i64; 2]; n];
        fn dfs(i: usize, j: usize, nums: &[i32], f: &mut [[i64; 2]]) -> i64 {
            if i >= nums.len() - 1 {
                return 0;
            }
            if f[i][j] != -1 {
                return f[i][j];
            }
            let cost = 0.max(nums[i - 1].max(nums[i + 1]) + 1 - nums[i]);
            let mut ans = cost as i64 + dfs(i + 2, j, nums, f);
            if j > 0 {
                ans = ans.min(dfs(i + 1, 0, nums, f));
            }
            f[i][j] = ans;
            ans
        }
        dfs(1, ((n & 1) ^ 1), &nums, &mut f)
    }
}
